src/main/java/org/nuxeo/ecm/core/storage/marklogic/CursorService.java

```diff
@@ -0,0 +1,228 @@
+/*
+ * (C) Copyright 2017 Nuxeo SA (http://nuxeo.com/) and others.
+ * 
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ * 
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *  
+ * Contributors:
+ *     Kevin Leturc
+ */
+package org.nuxeo.ecm.core.storage.marklogic;
+
+import static org.nuxeo.ecm.core.api.ScrollResultImpl.emptyResult;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.function.Function;
+
+import org.apache.commons.logging.Log;
+import org.apache.commons.logging.LogFactory;
+import org.nuxeo.ecm.core.api.NuxeoException;
+import org.nuxeo.ecm.core.api.ScrollResult;
+import org.nuxeo.ecm.core.api.ScrollResultImpl;
+
+/**
+ * A cursor service which holds cursors on DB in order to perform scroll operations.
+ *
+ * @param <C> The cursor type.
+ * @param <O> The cursor item type.
+ * @since 9.1
+ */
+public class CursorService<C, O> {
+
+    private static final Log log = LogFactory.getLog(CursorService.class);
+
+    protected Map<String, CursorResult<C, O>> cursorResults = new ConcurrentHashMap<>();
+
+    public void checkForTimedOutScroll() {
+        cursorResults.entrySet().stream().forEach(e -> isScrollTimedOut(e.getKey(), e.getValue()));
+    }
+
+    protected boolean isScrollTimedOut(String scrollId, CursorResult<C, O> cursorResult) {
+        if (cursorResult == null || cursorResult.timedOut()) {
```

Comment:
`ConcurrentHashMap` doesn't allow `null` values so `cursorResult == null` will never be true.